di/main.py

```diff
@@ -112,11 +112,20 @@ def __init__(self, config=ConfigManager):
 
     def wrapper(fn, warn=True):
         # Extract default values for keyword arguments
-        args, varargs, keywords, defaults = inspect.getargspec(fn)
-        if defaults:
-            defaults = dict(zip(reversed(args), reversed(defaults)))
-        else:
-            defaults = {}
+        try:
+            args, _ , _ , defaults = inspect.getargspec(fn)[:4]
+            if defaults:
+                defaults = dict(zip(reversed(args), reversed(defaults)))
+            else:
+                defaults = {}
+        except ValueError:
+            if not PY2:
+                # try to got it from kwonlydefaults as a sort of functions
+                # with a signature such as fn(*args, di=di_value, **kwargs)
+                # defaults come well packed
+                defaults = inspect.getfullargspec(fn)[5]
+                if not defaults:
+                    defaults = {}
```

Comment:
thanks @pfreixes! sorry for the late reply, didn't catch up very quickly with the personal mail after holidays 😳 

I've been thinking for quite a while about making use of new PY3 features in the library, namely supporting type annotations, so totally aligned in making this change, however I don't think the above logic covers all the scenarios.

Take the following example:

```py
>>> def foo(x, y='Y', *args, z='Z', **kwargs): pass
>>> inspect.getfullargspec(foo)
FullArgSpec(args=['x', 'y'], varargs='args', varkw='kwargs', defaults=('Y',), kwonlyargs=['z'], kwonlydefaults={'z': 'Z'}, annotations={})
```

unfortunately that forces us to merge `args` + `defaults` and `kwonlydefaults`. It seems doable but it'll require some solid test harness to ensure we're covering all the cases. There is now also the `inspect.signature` that offers a cleaner interface to the parameters. What do you think about going with a logic like the following? (:warning: untested code)

```py
try: # PY35
  sign = inspect.signature(fn, follow_wrapped=False)
  defaults = {
    p.name: p.default
    for p in sign.parameters.values()
    if p.default is not p.empty
  }
except (TypeError, ValueError, AttributeError):
  try: # PY3
    args, _, _, defaults, _, kwonlydefaults, _ = inspect.getfullargspec(fn)
  except AttributeError: # PY2
    args, _, _, defaults = inspect.getargspec(fn)
    kwonlydefaults = None

  defaults = dict(zip(reversed(args), reversed(defaults))) if defaults else {}
  if kwonlydefaults:
    defaults.update(kwonlydefaults)
``` 

with a bit more work it should support *annotations* too, which in PY3-only code bases I guess would be the desired way to define dependencies.

Do you feel like having a go at it?